apps/server/src/modules/room/api/room.uc.ts

```diff
@@ -162,30 +154,108 @@ export class RoomUc {
 		roleName: RoleName
 	): Promise<void> {
 		this.checkFeatureEnabled();
-		const roomAuthorizable = await this.checkRoomAuthorization(currentUserId, roomId, Action.write, [
+		const roomAuthorizable = await this.checkRoomAuthorizationByIds(currentUserId, roomId, Action.write, [
 			Permission.ROOM_MEMBERS_CHANGE_ROLE,
 		]);
 		this.preventChangingOwnersRole(roomAuthorizable, userIds, currentUserId);
 		await this.roomMembershipService.changeRoleOfRoomMembers(roomId, userIds, roleName);
-		return Promise.resolve();
 	}
 
 	public async passOwnership(currentUserId: EntityId, roomId: EntityId, targetUserId: EntityId): Promise<void> {
 		this.checkFeatureEnabled();
-		const roomAuthorizable = await this.checkRoomAuthorization(currentUserId, roomId, Action.write, [
-			Permission.ROOM_CHANGE_OWNER,
+		const ownershipContext = await this.getPassOwnershipContext(roomId, currentUserId, targetUserId);
+
+		this.checkRoomAuthorizationByContext(ownershipContext, Action.write, [Permission.ROOM_CHANGE_OWNER]);
+		this.checkUserInRoom(ownershipContext);
+		this.checkUserIsStudent(ownershipContext);
+
+		await this.roomMembershipService.changeRoleOfRoomMembers(roomId, [currentUserId], RoleName.ROOMADMIN);
+		await this.roomMembershipService.changeRoleOfRoomMembers(roomId, [targetUserId], RoleName.ROOMOWNER);
+	}
+
+	public async leaveRoom(currentUserId: EntityId, roomId: EntityId): Promise<void> {
+		this.checkFeatureEnabled();
+		await this.checkRoomAuthorizationByIds(currentUserId, roomId, Action.read, [Permission.ROOM_LEAVE]);
+		await this.roomMembershipService.removeMembersFromRoom(roomId, [currentUserId]);
+	}
+
+	public async removeMembersFromRoom(currentUserId: EntityId, roomId: EntityId, userIds: EntityId[]): Promise<void> {
+		this.checkFeatureEnabled();
+		await this.checkRoomAuthorizationByIds(currentUserId, roomId, Action.write, [Permission.ROOM_MEMBERS_REMOVE]);
+		await this.roomMembershipService.removeMembersFromRoom(roomId, userIds);
+	}
+
+	private checkFeatureEnabled(): void {
+		if (!this.configService.get('FEATURE_ROOMS_ENABLED', { infer: true })) {
+			throw new FeatureDisabledLoggableException('FEATURE_ROOMS_ENABLED');
+		}
+	}
+
+	private async getPassOwnershipContext(
+		roomId: EntityId,
+		currentUserId: EntityId,
+		targetUserId: EntityId
+	): Promise<OwnershipContext> {
+		const [targetUser, xyzContext] = await Promise.all([
+			this.userService.findById(targetUserId),
+			this.getRoomContext(roomId, currentUserId),
+		]);
+		const context = { ...xyzContext, targetUser };
```

Comment:
suggestion: renaming xyzContext to roomContext and keeping the order for better readability
```suggestion
		const [roomContext, targetUser] = await Promise.all([
			this.getRoomContext(roomId, currentUserId),
			this.userService.findById(targetUserId),
		]);
		const context = { ...roomContext, targetUser };
```